Incorporate DSL2 features into docs (#3793)
nextflow-io/nextflow: docs/workflow.rst

.. _workflow-page:

*********
Workflows
*********

> .. note::

Reviewer: As with modules, my preference would be to drop this down below the intro sentence.

    Workflows were introduced in DSL2. If you are still using DSL1, see the :ref:`dsl1-page` page to
    learn how to migrate your Nextflow pipelines to DSL2.

In Nextflow, a **workflow** is a composition of processes and dataflow logic (i.e. channels and operators).

The workflow definition starts with the keyword ``workflow``, followed by an optional name, and finally the workflow body
delimited by curly braces. A basic workflow looks like the following example::

    workflow {
        foo()
    }

Where ``foo`` could be a function, a process, or another workflow.

Workflows are *lazily executed*, which means that Nextflow parses the entire workflow structure first, and then
executes the entire workflow at once. The order in which a task is executed is determined only by its dependencies, so a task
will be executed as soon as all of its required inputs are available.

The syntax of a workflow is defined as follows::

    workflow [ name ] {

        take:
        < workflow inputs >

        main:
        < dataflow statements >

        emit:
        < workflow outputs >

    }

> .. note::

Reviewer: ```suggestion
.. tip::
```

Author: :+1:

    The ``main:`` label can be omitted if there are no ``take:`` or ``emit:`` blocks.

Invoking processes
==================

A process can be invoked like a function in a workflow definition, passing the expected
input channels like function arguments. For example::

    process foo {
        output:
        path 'foo.txt'

        script:
        """
        your_command > foo.txt
        """
    }

    process bar {
        input:
        path x

        output:
        path 'bar.txt'

        script:
        """
        another_command $x > bar.txt
        """
    }

    workflow {
        data = channel.fromPath('/some/path/*.txt')
        foo()
        bar(data)
    }

.. warning::

>     A process can be invoked only once in the same workflow. See :ref:`module-aliases` for
>     a workaround.

Reviewer: ```suggestion
    A process can be only be invoked once in a single workflow, unless using :ref:`module-aliases`.
```
